felix/nftables/maps.go

```diff
@@ -0,0 +1,677 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nftables
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/knftables"
+
+	"github.com/projectcalico/calico/felix/deltatracker"
+	"github.com/projectcalico/calico/felix/ipsets"
+	"github.com/projectcalico/calico/felix/logutils"
+	"github.com/projectcalico/calico/libcalico-go/lib/set"
+)
+
+var gaugeVecNumMaps = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Name: "felix_nft_maps",
+	Help: "Number of active Calico nftables maps.",
+}, []string{"ip_version"})
+
+type MapType string
+
+const MapTypeInterfaceMatch MapType = "interfaceMatch"
+
+type MapsDataplane interface {
+	AddOrReplaceMap(meta MapMetadata, members map[string][]string)
+	RemoveMap(setID string)
+
+	MapUpdates() *mapUpdates
+	FinishMapUpdates(updates *mapUpdates)
+	LoadDataplaneState() error
+}
+
+var _ MapsDataplane = &Maps{}
+
+type MapMetadata struct {
+	ID   string
+	Type MapType
+}
+
+// Maps manages a whole "plane" of maps, i.e. all the IPv4 maps, or all the IPv6 maps.
+type Maps struct {
+	IPVersionConfig *ipsets.IPVersionConfig
+
+	// mapNameToAllMetadata contains an entry for each map that has been
+	// added by a call to AddOrReplaceMap (and not subsequently removed).
+	mapNameToAllMetadata map[string]MapMetadata
+
+	// mapNameToProgrammedMetadata tracks the maps that we want to program and
+	// those that are actually in the dataplane.  It's Desired() map is the
+	// subset of mapNameToAllMetadata.
+	// Its Dataplane() map contains all maps matching the IPVersionConfig
+	// that we think are in the dataplane.  This includes any temporary IP
```

Comment:
```suggestion
	// that we think are in the dataplane.  This includes any temporary
```
Few bits of IP set terminology left over?